fuels-contract/src/contract.rs

```diff
@@ -78,44 +78,55 @@ impl Contract {
+        let script_len = 56;
         let script_data_offset = VM_TX_MEMORY + Transaction::script_offset() + script_len;
         let script_data_offset = script_data_offset as Immediate12;
 
+        // The offset that locates the asset ID in the script data.
+        // It goes from the beginning of `script_data` to `32`.
+        let asset_id_offset = 32;
+
         // Script to call the contract. The offset that points to the `script_data` is loaded at the
-        // register `0x10`. Note that we're picking `0x10` simply because it could be any
+        // register `0x12`. Note that we're picking `0x12` simply because it could be any
         // non-reserved register. Then, we use the Opcode to call a contract: `CALL` pointing at the
         // register that we loaded the `script_data` at.
-
-        // The `iter().collect()` does the Opcode->u8 conversion
         #[allow(clippy::iter_cloned_collect)]
         let script = vec![
-            Opcode::ADDI(0x10, REG_ZERO, script_data_offset),
-            // @todo currently there's no way to forward an amount.
-            // This would be done by programmatically changing
-            // $rB (`REG_ZERO` right now) to actually point to an
-            // amount. This would forward the amount in $rB using
-            // $rC as the asset_id.
-            Opcode::CALL(0x10, REG_ZERO, 0x10, REG_CGAS),
+            // Setting `0x10` to point to the 32-byte asset ID of the amount to forward.
+            Opcode::ADDI(0x10, REG_ZERO, asset_id_offset),
+            // Setting `0x11` to hold the amount of of coins to forward.
+            Opcode::ADDI(0x11, REG_ZERO, tx_parameters.amount as Immediate12),
+            // Setting `0x12` to the `script_data`, defined down below but
+            // we already know its length.
+            Opcode::ADDI(0x12, REG_ZERO, script_data_offset),
+            Opcode::CALL(0x12, 0x11, 0x10, REG_CGAS),
             Opcode::RET(REG_RET),
             Opcode::NOOP,
         ]
         .iter()
         .copied()
         .collect::<Vec<u8>>();
 
-        assert_eq!(script.len(), script_len, "Script length *must* be 16");
+        assert_eq!(
+            script.len(),
+            script_len - asset_id_offset as usize,
+            "Script length *must* be 24"
+        );
 
         // `script_data` consists of:
-        // 1. Contract ID (ContractID::LEN);
-        // 2. Function selector (1 * WORD_SIZE);
-        // 3. Calldata offset, if it has structs as input,
+        // 1. Asset ID to be forwarded
+        // 2. Contract ID (ContractID::LEN);
+        // 3. Function selector (1 * WORD_SIZE);
+        // 4. Calldata offset, if it has structs as input,
         // computed as `script_data_offset` + ContractId::LEN
         //                                  + 2 * WORD_SIZE;
-        // 4. Encoded arguments.
+        // 5. Encoded arguments.
         let mut script_data: Vec<u8> = vec![];
 
+        script_data.extend(tx_parameters.asset_id.to_vec());
```

Comment:
@luizstacio @AlicanC This whole section here is what's needed on `fuels-ts` to implement the same thing. Let me know if it's not clear enough and/or if you have more questions!